### ISSUE TITLE
beamer.cls support

### ISSUE THREAD

**Issue by kohlhase** [brucemiller/LaTeXML#231]
[Originally Ticket 1555]

We have made some progress on presenting slides in `slides.cls.ltxml` in #230. Based on this, we can start implementing LaTeXML bindings for `beamer.cls`, which is much more widely used. As the functionality of `beamer.cls` is a superset of that of `slides.cls`, this should be based on that. 

I have started a stub implementation on the `arXMLiv` branch to get a specific example (for the IMSSS) going. I will see how far I get with this and report on this ticket.

**Comment by kohlhase** [comment]
I would like to get this ticket to Bruce's attention. Even though it does not appear in the arXiv, beamer is quite widely used. And slides are an area, where people could really profit from LaTeXML support, as there are quite a few nice web-based slides with CSS.

**Comment by kohlhase** [comment]
The case to make here is that Beamer is quite semantic already, and that (even though beamer itself is a huge beast, a binding would be relatively doable. 

I think even relatively basic support for ```beamer.cls``` would be quite helpful. 
* the ```frame``` environment and macro as well as  ```frametitle``` and ```framesubtitle``` have quite natural XML counterparts, I think we probably have to extend the LTXML schema slightly there, I guess ```LaTeXML-slides.rnc``` would be very simple.  
* the theme stuff can just be passed on to the XML symbolically, then others can work on CSS and JS to realize these in the interface. LaTeXML does not have to do anything here. 
* the overlay specifications are more interesting, but largely a parsing issue. In a nutshell: ```beamer.cls``` allows overlay qualifiers to some macros like ```\item```, where ```\item<1,3->``` means that the item is shown in overlays 1 and 3ff. If I remember correctly @dginev already did some parsing extensions for sTeX once (I am not sure where they are). Again, these overlay specifications only have to be passed on to the XML. The presentation can then be left to suitable translations into web-slide packages. 
* there are a couple of overlay environments and macros ```\only```, ```\onlyenv```, ```\uncover```, ... they are relatively easy to add, again passing on the information into the XML. 

The macros above are (I believe) what most people are using. They would go a really long way along the 80/20 route.

**Comment by kohlhase** [comment]
Actually, I am seeing that there is quite some work by Deyan in this direction already as part of https://github.com/KWARC/LaTeXML-Plugin-sTeX/blob/master/lib/LaTeXML/Package/mikoslides.sty.ltxml
We could start extracting this into a ```beamer.cls.ltxml``` to start things off. 
Actually, maybe the right thing to do here is to start another plugin for LaTeXML?

**Comment by dginev** [comment]
Pushing this back to 0.8.4, since we should really just roll out 0.8.3 asap and continue adding enhancements after.

**Comment by krivit** [comment]
As a possible stopgap, what about implementing `beamerarticle.sty.ltxml` instead? That style mainly makes sure that `beamer` macros are defined, usually to pass their input through unchanged.

**Comment by dginev** [comment]
That indeed sounds like an easy add-on to the issue, and a much easier first deliverable.

**Comment by pksweby** [comment]
I'd like to add my support to the request for this. At present I have a series of LaTeX definitions to enable me to strip out the transition features of Beamer to produce an html version using LaTeXML, without having to destroy the original Beamer coding which is still used in parallel

**Comment by Jazzpirate** [comment]
+1

**Comment by kohlhase** [comment]
Is there any progress on this? Should basic TeX primitives support in LaTeXML not be sufficient to just read `beamer.cls` now that it can essentially do `tikz` and `pgf`?

**Comment by dginev** [comment]
Before I forget, here is another recent [beamer support request](https://twitter.com/reset_shift/status/1409648405287104512) addressed to a number of converters by name - pandoc, latexml, tex4ht, plastex. 

The feedback I got when I asked about the HTML target was "any of the pandoc-supported slide dialects would be great" for that specific user. Which, [according to the pandoc manual](https://pandoc.org/MANUAL.html#slide-shows), are five choices: S5, DZSlides, Slidy, Slideous, or reveal.js.

Their overall slide support is as usual based around a specialized pandoc AST that targets slides. One wonders whether (and how much) we  should or shouldn't use the pandoc implementations as a guide for the representation side of things, given they're a bit ahead of the game with limited support for various slideshow frameworks.

**Comment by nxg** [comment]
I'd like to add a +1 to @krivit's suggestion that it would be separately useful (and possibly more straightforwardly achievable) if LaTeXML were able to support beamer's article output mode.

Use case: I've been experimenting with running a colleague's lecture notes through LaTeXML.  They use `beamerswitch` (as illustrated below).  I'm interested in producing XHTML for EPUB, and for our purposes, if we were able to produce EPUB from the beamer article mode, that would tick quite a few of our (accessibility-related) boxes.  That might represent an intermediate goal, which could be achieved on the way towards the larger goal of converting beamer input to XHTML for presentation purposes.

At present (LaTeXML 0.8.5), LaTeXML complains with

    Error:undefined:\PassOptionsToClass The token T_CS[\PassOptionsToClass] is not defined.
    ...
    Warning:missing_file:beamerswitch Can't find binding for class beamerswitch (using OmniBus)

then does quite a lot of work, but is clearly rather confused.  It does produce output and a zero exit code, but the `.lxml` file is almost entirely empty.

How beamer article mode works (for reference):

    \PassOptionsToClass{a4paper,11pt}{article} % for article mode
    \PassOptionsToClass{smaller}{beamer} % for beamer, handout, trans modes
    \documentclass{beamerswitch}
    ....

...from which we can produce an article version with `xelatex -jobname=filename-article filename.tex`

**Comment by tkw1536** [comment]
Just to have it documented: Initial support for `beamer.cls` was merged in #1652.

**Comment by dginev** [comment]
@tkw1536 Indeed! 

We should try to find a nice HTML+JS target next, implement for it, and get a couple of themes, to show off to the world. Beamer is kind of tangential to my own interests, but it tends to be a very intriguing marketing lever for the wider LaTeX-using academic world.

**Comment by tkw1536** [comment]
True, that should be a next goal, in the past you suggested #1828  as a useful test case.

**Comment by brucemiller** [comment]
Since the basic beamer support is finally there, how about I close this issue (particularly since it's 11 years old, sigh).  And if you want a nice JS demo (which *would* be cool), you open a new issue, or better yet, a PR :>

**Comment by Kreijstal** [comment]
> Since the basic beamer support is finally there, how about I close this issue (particularly since it's 11 years old, sigh). And if you want a nice JS demo (which _would_ be cool), you open a new issue, or better yet, a PR :>

that would be really cool

**Comment by dginev** [comment]
Feel free to reuse the still open #1828 for any HTML+JavaScript discussions.
